src/webgpu/compat/api/validation/pipeline_creation.spec.ts

```diff
@@ -144,3 +144,206 @@ g.test('depth_textures')
         break;
     }
   });
+
+function numCombosToNumber(maxCombos: number, numCombos: string): number {
+  const m = /^max([+-]?)(\d+)$/.exec(numCombos);
+  if (m) {
+    if (m[1] === '+') {
+      return maxCombos + parseInt(m[2]);
+    } else if (m[1] === '-') {
+      return maxCombos - parseInt(m[2]);
+    } else {
+      return maxCombos;
+    }
+  }
+  return parseInt(numCombos);
+}
+
+function numNonSampledToNumber(maxTextures: number, numNonSampled: '0' | '1' | 'max') {
+  switch (numNonSampled) {
+    case '0':
+      return 0;
+    case '1':
+      return 1;
+    case 'max':
+      return maxTextures;
+  }
+}
+
+g.test('texture_sampler_combos')
+  .desc(
+    `
+Test that you can not use more texture+sampler combos than
+min(maxSamplersPerShaderStage, maxSampledTexturesPerShaderStage)
+in compatibility mode.
+`
+  )
+  .params(u =>
+    // prettier-ignore
+    u
+      .combineWithParams([
+        ...([
+          { pass: true,  numCombos: 'max',   numNonSampled: '1',   numExternal: 0, useSame: false },
+          { pass: false, numCombos: 'max+1', numNonSampled: '1',   numExternal: 0, useSame: false },
+          { pass: true,  numCombos: '1',     numNonSampled: 'max', numExternal: 0, useSame: false },
+          { pass: false, numCombos: '2',     numNonSampled: 'max', numExternal: 0, useSame: false },
+          { pass: true,  numCombos: 'max-4', numNonSampled: '0',   numExternal: 1, useSame: false },
+          { pass: false, numCombos: 'max-3', numNonSampled: '0',   numExternal: 1, useSame: false },
+          { pass: true,  numCombos: 'max-8', numNonSampled: '0',   numExternal: 2, useSame: false },
+          { pass: false, numCombos: 'max-7', numNonSampled: '0',   numExternal: 2, useSame: false },
+          { pass: true,  numCombos: 'max-7', numNonSampled: '0',   numExternal: 2, useSame: true },
+          { pass: false, numCombos: 'max-6', numNonSampled: '0',   numExternal: 2, useSame: true },
```

Comment:
nit: it'd be good to have comments or language in description explaining each external texture/sampler takes 3 slot (YUV) IIUC

Reply:
I don't mind adding comments but is this a pattern from other tests? Can you give an example of what you want to see? I can copy and paste the spec but that seems wrong given the spec might change and I don't see the spec being copy and pasted in other tests. To see what's being tested shouldn't you just reference the spec? I started to write a comment per line but after writing 5 I realized I was just repeating the data in the params.

Reply:
Okay then let's just leave it as is. Future contributors can just rely on spec.
(Context: I was a bit confused on external texture takes 3 on first glance so I made this comment.)